pkg/apis/k3s.cattle.io/v1/types.go

```diff
@@ -1,20 +1,97 @@
 package v1
 
 import (
+	"k8s.io/apimachinery/pkg/api/resource"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
 // +genclient
+// +genclient:noStatus
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
+// Addon is used to track application of a manifest file on disk. It mostly exists so that the wrangler DesiredSet
+// Apply controller has an object to track as the owner, and ensure that all created resources are tracked when the
+// manifest is modified or removed.
 type Addon struct {
 	metav1.TypeMeta   `json:",inline"`
 	metav1.ObjectMeta `json:"metadata,omitempty"`
 
+	// Spec provides information about the on-disk manifest backing this resource.
 	Spec AddonSpec `json:"spec,omitempty"`
 }
 
 type AddonSpec struct {
-	Source   string `json:"source,omitempty"`
-	Checksum string `json:"checksum,omitempty"`
+	// Source is the Path on disk to the manifest file that this Addon tracks.
+	Source string `json:"source,omitempty" column:""`
+	// Checksum is the SHA256 checksum of the most recently successfully applied manifest file.
+	Checksum string `json:"checksum,omitempty" column:""`
+}
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ETCDSnapshot tracks a point-in-time snapshot of the etcd datastore.
+type ETCDSnapshotFile struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines properties of an etcd snapshot file
+	Spec ETCDSnapshotSpec `json:"spec,omitempty"`
+	// Status represents current information about a snapshot.
+	Status ETCDSnapshotStatus `json:"status,omitempty"`
+}
+
+// ETCDSnapshotSpec desribes an etcd snapshot file
+type ETCDSnapshotSpec struct {
+	// NodeName contains the name of the node on which the snapshot file resides.
+	// For snapshots uploaded to S3, the name will be "s3".
+	NodeName string `json:"nodeName,omitempty" column:"name=Node"`
+	// Location is the on-disk path or S3 URI of the snapshot file.
+	Location string `json:"location,omitempty" column:""`
+	// Metadata contains point-in-time snapshot of the contents of the k3s-etcd-snapshot-extra-metadata ConfigMap's data field, at the time the snapshot was taken.
```

Comment:
There was a request by @Oats87 for keyHash of the token in use when the snapshot was taken to be saved in the metadata. Maybe I missed it somewhere in all the changes, but has this data been added to the snapshotspec?

Reply:
Thanks for the reminder. I forgot to add that to the ADR since I think it came up in person. The current metadata is directly copied from the ConfigMap, and I don't think it belongs in a field on the snapshot itself; I think I'll keep it in an annotation on the resource.